RollbarNotifier/Sources/RollbarNotifier/RollbarSession.m

```diff
@@ -0,0 +1,434 @@
+    // self->_state.appID = NOTE: we do not want to ever override this value...
+    self->_state.sessionID = [NSUUID new];
+    self->_state.sessionStartTimestamp = [NSDate date];
+    self->_state.appMemoryWarningTimestamp = nil;
+    self->_state.appTerminationTimestamp = nil;
+    self->_state.sysSignal = nil;
+    self->_state.appCrashDetails = nil;
+    self->_state.appInBackgroundFlag = RollbarTriStateFlag_None;
+    
+    [self saveCurrentSessionState];
+    
+    [self registerForSystemSignals];
+    [self registerApplicationHooks];
+}
+
+- (void)deduceOomTermination {
+    
+    RollbarSdkLog(@"Deducing OOOM Termination based on: %@", self->_state);
+    
+    BOOL appQuit = NO;
+    struct stat statbuffer;
+    if (stat(appQuitFilePath, &statbuffer) == 0){
+        // A file exists at the path, we had an intentional quit
+        appQuit = YES;
+    }
+    if (YES == appQuit) {
+
+        NSString *message = @"The app was exited at the end of last session.";
+        [Rollbar infoMessage:message data:self->_state.jsonFriendlyData];
+        return;
+    }
+    
+    BOOL appUpgraded = [self didAppVersionChange:self->_state.appVersion];
+    if (YES == appUpgraded) {
+        
+        NSString *message = [NSString stringWithFormat:@"The app was upgraded from %@ to %@",
+                             self->_state.appVersion,
+                             [RollbarBundleUtil detectAppBundleVersion]
+        ];
+        [Rollbar infoMessage:message data:self->_state.jsonFriendlyData];
+        return;
+    }
+    
+    BOOL appTerminated = (nil != self->_state.appTerminationTimestamp) ? YES : NO;
+    if (YES == appTerminated) {
+        
+        NSString *message = [NSString stringWithFormat:@"The app was terminated at %@",
+                             self->_state.appTerminationTimestamp
+        ];
+        RollbarSdkLog(@"%@", message);
+        [Rollbar infoMessage:message data:self->_state.jsonFriendlyData];
+        return;
+    }
+    
+    BOOL appCrashed = self->_crashLocator();
+    if (YES == appCrashed) {
+        
+        NSString *message = @"The app had a crash.";
+        RollbarSdkLog(@"%@", message);
+        [Rollbar criticalMessage:message data:self->_state.jsonFriendlyData context:self->_state.appCrashDetails];
+        return;
+    }
+    
+    BOOL osUpgraded = [self didOsVersionChange:self->_state.osVersion];
+    if (YES == osUpgraded) {
+        
+        NSString *message = [NSString stringWithFormat:@"The OS was upgraded from %@ to %@",
+                             self->_state.osVersion,
+                             [RollbarOsUtil detectOsVersionString]
+        ];
+        RollbarSdkLog(@"%@", message);
+        [Rollbar infoMessage:message data:self->_state.jsonFriendlyData];
+        return;
+    }
+    
+    NSString *oomContext = nil;
+    switch (self->_state.appInBackgroundFlag) {
+            
+        case RollbarTriStateFlag_On:
+            oomContext = @"BOOM";
```

Comment:
Would we get here in the vent the app is simply shut down by the OS because it's hasn't come to foreground for too long, or because the current foreground app requires memory and the OS is just terminating background apps to free memory?

Reply:
We only get here if concluded that this is a possible OOM case so we also add FOOM or BOOM classification detail.
What i see based on my testing is that there will be false positive OOM detections (as we suspected).
The good news is that we have some other data available that we can start using in order to minimize the false positives.

Reply:
I'm mostly concerned that BOOM reports may be entirely composed of the very common event that is the OS simply freeing memory for foreground apps. Which could result in over-reporting OOM problems that are not exactly a problem.

I wonder if it'd be best to report either at least with a separate message. Thoughts?

Reply:
@matux , that is an absolutely valid concern and one of the reasons to make this new OOM detection optional and to be able to get configured to stay OFF. Also, in the soon-upcoming development iteration, we want to couple current decision-making about possible BOOM/FOOMs with taking into account memory stats tracking that we already have.